### DISCUSSION TITLE
Migrate from Microsoft 365 to Another IMAP it's not easy anymore but it can be done [procedures]

### DISCUSSION THREAD

**Discussion by mshannaq** [imapsync/imapsync#435] [Show and tell]
As Imapsync says in docs:

**Can I use imapsync to transfer from or to Office365 accounts?**

**Yes, but it's not easy anymore for new Azure accounts.**

``` text
For now, authentication to Office365 with Imapsync is not a piece of
cake (June 2023).

One authentication method is the LOGIN method, the second is the app
password and the third is the OAUTH2 method.
```

So you have to do 3 steps 🥹😅

Today I spent  a couple of hours to do it , and finally it works , and I would like to share the procedures  that I did:

Read it all before try...

First of all you have to create app password for the account you want to migrate (see https://ae.godaddy.com/help/create-app-passwords-40980)

and to do that you have to make sure you enable multi-factor authentication (see https://ae.godaddy.com/help/enable-or-disable-multi-factor-authentication-32039) and make sute that you enable force multi-factor authentication for the users you want to migrate from (https://account.activedirectory.windowsazure.com/UserManagement/Multifactorverification.aspx#) (need you to sign in with your microsoft 356 account)

then  after that return to your account (on https://mysignins.microsoft.com/security-info ) and create app password for microsoft 365 account (and keep the password with you in safe please because it is shown once).

![image](https://github.com/imapsync/imapsync/assets/162535/4a730c8b-3a9d-45e0-845a-f3270b75d478)
**note**: you will not see the "App Password" Choice until you activate multi-factor authentication for that user.

after that you can use `imapsync` so it have to be installed on your machine (see https://imapsync.lamiral.info/#install) (on my mac i use `brew install imapsync`)

but first you have to generate oauth2 token as described in https://imapsync.lamiral.info/oauth2/oauth2_office365/README.txt by

```
wget https://imapsync.lamiral.info/oauth2/oauth2_office365/oauth2_office365_with_imap
```

then run `oauth2_office365_with_imap` as

on mac i use:
```
./oauth2_office365_with_imap email-you-want-to-migrate@yourdomain.com
```
this will open a browser and ask you to sign in and generate oauth2 token , accept the request to continue.

This will create token file, The tokens will printed on the screen and also  generated in the sub-directory "tokens" which has to
exist before contains 2 lines , one for the `access token` and the second for `refresh token`

example:

```
./imapsync --host1 outlook.office365.com  --user1 email-you-want-to-migrate@yourdomain.com  --oauthaccesstoken1  /root/tokens/oauth2_tokens_email@domain.txt  --host2 imap.newserver --user2 newimap@you.com  --password2  "YourImapPasswprd"   --automap --justlogin
```

This will try to login on each servers, if everything fine then remove `--justlogin` from the command and run it again to start migrate (sync)

**Note:** when you try to run the command the terminal may ask you for the email password , enter the **app password** that you create for the email in above steps.

and destination email account should be exists and accessable  via IMAP  also

**(Updates) Note that :** For emails that does not has 2fa enabled I notices that I can just create oauth2 token only and sync its contents without need to enforce it to use app password. 😅

Thanks

## Resources that help me much more:
* https://imapsync.lamiral.info/FAQ.d/FAQ.Office365.txt
* https://imapsync.lamiral.info/oauth2/oauth2_office365/README.txt

**Comment by gilleslamiral** [comment]
> So you have to do 3 steps

Since you ended up having to do OAUTH2, I don't understand why you didn't start doing OAUTH2 and only it.

**Reply by mshannaq** [reply]
that what i discovered lately 🥹

but also I see that while migration account that contains big email count for example 500K email messages for example , it take too long to migrate and fial also. even if I refresh the token every 30 minutes . it just stuck and do nothing.

**Reply by gilleslamiral** [reply]
Which imapsync release do you use?

https://imapsync.lamiral.info/ChangeLog
````...
revision 2.230
date: 2023/01/24 14:36:04;  author: gilles;  state: Exp;  lines: +190 -119
Bugfix. Reread oauth2 files upon disconnection
````
Latest: https://imapsync.lamiral.info/imapsync

**Comment by David-Garcia9919** [comment]
Nice explanation — getting OAuth2 working with ImapSync definitely takes some patience.

For anyone who needs an easier setup, EdbMails [Office 365 Migration](https://www.edbmails.com/pages/office-365-migration-software.html) tool can automate the mailbox transfer and manage OAuth2 tokens internally. It’s not open source, but it’s handy when you just want a direct, script-free migration process.

**Comment by krazzer** [comment]
It took me hours to make this work, so I decided to make this little guide to save someone else some time:

# IMAP Migration with OAuth2 (Exchange Online → IMAP)

This guide describes how to migrate emails from Microsoft Exchange Online to a standard IMAP server using `imapsync` and OAuth2 authentication.

---

## 1. Login to Azure Portal
Go to [https://portal.azure.com](https://portal.azure.com) with your Exchange account.

---

## 2. Register a new application
1. In the left menu, under **Azure Services**, click **App registrations**.  
2. Click **New registration**.  
3. Name the app `imapsync` and click **Register**.  
4. Save the **Application (client) ID** shown.

---

## 3. Configure Authentication
1. In the app, click **Authentication** in the left menu.  
2. Under **Supported account types**, select:  
   `Accounts in any organizational directory (Any Microsoft Entra ID directory – Multitenant)`  
3. Scroll to **Advanced settings / Allow public client flows**, and **enable** mobile & desktop flows.  
4. Click **Add a platform** → **Mobile and desktop applications**.  
5. Under **Redirect URIs**, enter:  `https://localhost`
6. Click **Configure**, then **Save**.

---

## 4. Configure API Permissions
1. Click **API permissions** in the left menu.  
2. Click **Add a permission** → **Microsoft Graph** → **Delegated permissions**.  
3. Search for `IMAP`, expand **IMAP**, and check `IMAP.AccessAsUser.All`.  
4. Click **Add permissions**.

---

## 5. Prepare `oauth2_imap`
1. Download [oauth2_imap.zip](https://imapsync.lamiral.info/oauth2/oauth2_imap.zip) and unzip.

---

## 6. Generate OAuth2 token
Run the command in the terminal (replace placeholders with your actual values):

```bash
./oauth2_imap/oauth2_imap \
--provider office365
--client_id "CLIENT_ID" \
--token_file "EMAIL_ADDRESS.token.txt" \
EMAIL_ADDRESS
```

## 7. Run IMAP sync
After generating the OAuth2 token, execute:
```bash
imapsync \
  --host1 outlook.office365.com \
  --user1 "EMAIL_ADDRESS" \
  --oauthaccesstoken1 "EMAIL_ADDRESS.token.txt" \
  --host2 NEW_HOST_SERVER \
  --user2 "EMAIL_ADDRESS" \
  --password2 "NEW_PASSWORD" \
  --automap \
```

**Reply by gilleslamiral** [reply]
The oauth2_imap can be reduced to:
````
./oauth2_imap/oauth2_imap \
--provider office365
--client_id "CLIENT_ID" \
--token_file "EMAIL_ADDRESS.token.txt" \
EMAIL_ADDRESS
````
but the --redirect_uri value will be `https://localhost`, not `http://localhost`

From the imapsync command example, you can remove `--ssl1 --ssl2 --syncinternaldates --authmech1 XOAUTH2` as they are already set.

**Reply by krazzer** [reply]
Tnx! Updated.

**Reply by gr0esch** [reply]
Really helpful, thanks a lot!

**Comment by demetrioparrilla** [comment]
Migrating from Microsoft 365 to an IMAP account has become more complex due to modern authentication requirements and the need to preserve mailbox structure. While manual methods are available, they can be time-consuming, especially for large mailboxes or multiple user accounts.
If you're looking for a simpler approach, you can try the **Softaken Office 365 to IMAP Migration Tool**. It supports OAuth authentication, preserves the original folder hierarchy, and helps transfer emails efficiently from Microsoft 365 to IMAP-enabled email services. The interface are simple and easy, no tech skills are needed to use. I hope this solves your problem.

**Reply by nine** [reply]
This is a SPAM post.

**Reply by nine** [reply]
The profile image of @demetrioparrilla is used for many different identities on the internet. You can try a reverse image search on google search by yourself.

**Comment by techgeek-khushi-2029** [comment]
One thing I'd check first is whether the destination IMAP server enforces connection or upload throttling. A lot of migrations "look" like authentication problems but are actually rate limits kicking in after a few successful syncs.

If you don't want to script around those limitations, I've had good results with BitRecover Office 365 to IMAP Migration Wizard. It handles multiple mailboxes, preserves folder hierarchy, and retries transient failures automatically, which saves a lot of manual reruns.

Also, before any large migration, it's worth comparing the destination server's maximum folder depth and message size limits. That's something many people miss until folders start getting skipped.

If it helps, BitRecover also has a concise guide covering Office 365 to IMAP migration scenarios: https://www.bitrecover.com/blog/migrate-office-365-to-imap-server/

**Comment by techinmay** [comment]
Migrating from Microsoft 365 to another IMAP email service isn't as straightforward as it used to be because of modern authentication, security policies, and mailbox throttling, but it can still be done successfully. While manually configuring both accounts in an email client and copying emails is an option, it is time-consuming and risky for large mailboxes. A more reliable approach is to use the **SysTools [IMAP Migrator](https://www.systoolsgroup.com/imap/migration/)**, which enables direct migration from Microsoft 365 to IMAP-supported email servers while preserving folder hierarchy, email metadata, and attachments. It also supports batch mailbox migration, incremental migration to prevent duplicates, and detailed reporting, making it an efficient choice for both individual users and organizations.

**Reply by SidSHaresSoftware** [reply]
Thanks for this recommendation

**Reply by gilleslamiral** [reply]
The description above from Techinmay is also perfect to describe Imapsync! Thanks!

**Comment by shiv2expert-glitch** [comment]
The [Recoveryfix Email Migration ](https://www.recoveryfix.com/email-migration/)Tool is a professional software designed to securely migrate mailboxes between IMAP servers, cloud platforms, and Exchange environments. It simplifies transitions—like moving to Microsoft 365 or tenant-to-tenant migrations—with zero data loss. It features batch migration, folder hierarchy preservation, and smart filters for efficient, selective data transfers.
